### PR TITLE
vmware, network: add maclearning option

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Adds option for specifying MAC Learning property for network offering (useful for VMware Distributed Virtual Portgroup). Added global config - `network.mac.learning` for the default value.
MAC Learning is supported for DV portgroups for VMware Distributed vSwitches v6.6.0+ and vSphere 6.7+

Documentation PR: https://github.com/apache/cloudstack-documentation/pull/242

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Screenshot from 2021-08-05 16-17-50

### CODE REVIEW

**Comment by shwstppr** [comment]
Moving MAC learning addition changes from #5181 here

@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java [hunk 1]
```diff
@@ -839,10 +849,79 @@ private static void createPortGroup(String physicalNetwork, String networkName,
         }
     }
 
-    public static boolean isSpecMatch(DVPortgroupConfigInfo currentDvPortgroupInfo, DVPortgroupConfigSpec newDvPortGroupSpec) {
+    private static boolean eitherObjectNull(Object obj1, Object obj2) {
+        return (obj1 == null && obj2 != null) || (obj1 != null && obj2 == null);
+    }
+
+    private static boolean areBoolPoliciesDifferent(BoolPolicy currentPolicy, BoolPolicy newPolicy) {
+        return eitherObjectNull(currentPolicy, newPolicy) ||
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:857]: any case where both objs are null?

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java [hunk 2]
```diff
@@ -839,10 +849,79 @@ private static void createPortGroup(String physicalNetwork, String networkName,
...
+                (newPolicy != null && newPolicy.isValue() != currentPolicy.isValue());
```
Comment (sureshanaparti) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: extra not null checks may not be required (both objs are not null ensured before using _eitherObjectNull()_), please check wherever applicable.

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java (hunk 1 above)
Comment (shwstppr) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:857]: @sureshanaparti if both objects are null that would mean both policies are essentially same (`null`). Method should return false in that case

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java (hunk 2 above)
Comment (shwstppr) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: There is no check for both objects null. `eitherObjectNull()` checks if only one of them is null.

**Comment by shwstppr** [comment]
^^ could be an intermittent failure for tests:
```2021-09-20 23:06:06,436 - ERROR - marvinRequest : CmdName: <marvin.cloudstackAPI.createNetwork.createNetworkCmd object at 0x7fe468167898> Exception: ['Traceback (most recent call last):\n', '  File "/usr/local/lib/python3.6/site-packages/marvin/cloudstackConnection.py", line 376, in marvinRequest\n    raise self.__lastError\n', '  File "/usr/local/lib/python3.6/site-packages/marvin/cloudstackConnection.py", line 310, in __parseAndGetResponse\n    response_cls)\n', '  File "/usr/local/lib/python3.6/site-packages/marvin/jsonHelper.py", line 155, in getResultObj\n    raise cloudstackException.CloudstackAPIException(respname, errMsg)\n', 'marvin.cloudstackException.CloudstackAPIException: Execute cmd: createnetwork failed, due to: errorCode: 431, errorText:The VLAN tag to use for new guest network, 1201 is already being used for dynamic vlan allocation for the guest network in zone pr5471-t2136-vmware-67u3\n']
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/marvin/cloudstackConnection.py", line 376, in marvinRequest
    raise self.__lastError
  File "/usr/local/lib/python3.6/site-packages/marvin/cloudstackConnection.py", line 310, in __parseAndGetResponse
    response_cls)
  File "/usr/local/lib/python3.6/site-packages/marvin/jsonHelper.py", line 155, in getResultObj
    raise cloudstackException.CloudstackAPIException(respname, errMsg)
marvin.cloudstackException.CloudstackAPIException: Execute cmd: createnetwork failed, due to: errorCode: 431, errorText:The VLAN tag to use for new guest network, 1201 is already being used for dynamic vlan allocation for the guest network in zone pr5471-t2136-vmware-67u3
```
Will re-run
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java (hunk 2 above)
Comment (sureshanaparti) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: correct, so when both are null, can be a NPE?

Comment (shwstppr) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: @sureshanaparti 
Four cases:

1. currentPolicy is null, newPolicy is not null       <- Will be handled by `eitherObjectNull`
2. currentPolicy is not null, newPolicy is null       <- Will be handled by `eitherObjectNull`
3. currentPolicy is null, newPolicy is null             <- Will fail `newPolicy != null` check in second part of the condition
4. currentPolicy is not null, newPolicy is not null <- Should be handled by `newPolicy.isValue() != currentPolicy.isValue()`

Actually I added null check of both vars initially but IntelliJ suggested simplification and I guess it was right

Comment (sureshanaparti) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: @shwstppr why not you handle case 3) in the private method itself?

Comment (shwstppr) [reply vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/HypervisorHostHelper.java:858]: @sureshanaparti which private method?
If both objects are null, `areBoolPoliciesDifferent` method should return false. That is why

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@shwstppr @NuxRo is this ready for review?

**Comment by shwstppr** [comment]
@rhtyd should be ready this week, @alexandremattioli doing some tests for this and #5470

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Comment by rhtyd** [comment]
I see consistent new issue, can you investigate and fix @sureshanaparti @nvazquez cc @shwstppr

**Comment by rhtyd** [comment]
The vmware issue/failure exists on main branch as seen in #5201
Merging this

**Review by rhtyd — APPROVED** [review]

**Comment by rhtyd** [comment]
this is ready for merging but has conflicts due to https://github.com/apache/cloudstack/pull/5470/files cc @nvazquez @sureshanaparti @shwstppr

**Comment by sureshanaparti** [comment]
> this is ready for merging but has conflicts due to https://github.com/apache/cloudstack/pull/5470/files cc @nvazquez @sureshanaparti @shwstppr

Fixed conflicts (due to #5470 merging)

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test centos7 vmware-67u3

**Review by nvazquez — APPROVED** [review]

**Comment by nvazquez** [comment]
The test failure has been fixed by PR #5547, merging

**Pull request merged**